src/PaymentServer/Processors/Stripe.hs

```diff
@@ -83,3 +119,49 @@ webhook d Event{eventId=Just (EventId eventId), eventType=ChargeSucceededEvent,
 webhook d _ =
   -- TODO: Record the eventId somewhere.
   return Ok
+
+
+-- | Browser facing API that takes token, voucher and a few other information
+-- and calls stripe charges API. If payment succeeds, then the voucher is stored
+-- in the voucher database.
+type ChargesAPI = "charge" :> ReqBody '[JSON] Charges :> Post '[JSON] Acknowledgement
+
+data Charges = Charges
+  { token :: Text
+  , voucher :: Voucher
+  , amount :: Int
+  , currency :: Text
+  } deriving (Show, Eq)
+
+instance FromJSON Charges where
+  parseJSON (Object v) = Charges <$>
+                         v .: "token" <*>
+                         v .: "voucher" <*>
+                         v .: "amount" <*>
+                         v .: "currency"
+  parseJSON _ = mzero
+
+-- | call the stripe Charge API (with token, voucher in metadata, amount, currency etc
+-- and if the Charge is okay, then set the voucher as "paid" in the database.
+charge :: VoucherDatabase d => d -> ByteString -> Charges -> Handler Acknowledgement
+charge d key (Charges token voucher amount currency) = do
+  let config = StripeConfig (StripeKey key) Nothing
+      tokenId = TokenId token
+  result <- liftIO $ stripe config $
+    createCharge (Amount amount) (read (unpack currency))
+      -&- tokenId
+      -&- MetaData [("Voucher", voucher)]
+  case result of
+    Right Charge { chargeMetaData = metadata } ->
+      -- verify that we are getting the same metadata that we sent.
+      case metadata of
+        MetaData [("Voucher", v)] ->
+          if v == voucher
+            then
+            do
+              liftIO $ payForVoucher d voucher
+              return Ok
+            else
+            throwError err400 { errBody = "Voucher code mismatch" }
```

Comment:
Hm!  Do you have thoughts on particular scenarios where the voucher code will mismatch?  Is this something a user can provoke or is it an error in either our code or Stripe's server?  If the latter, I'd probably go with err500.

Reply:
What I had in mind was a scenario where for _some reason_, stripe didn't give us the metadata back. Sounds impossible. I changed it to return 500, but if you think this need not be handled, we can drop that ugly looking `if-then-else` from the function.

Reply:
It seems like a good thing to check.  An alternative might be to stop caring about the Stripe metadata and just associate a charge id with a voucher in the local database.  Then we have more local state to maintain but less reliance on Stripe.

I don't see a clear signing pointing one way or the other here so let's leave this as-is for now.  If we observe problems with this approach in testing or real-world usage we can revisit it.